lib/network_utils.ml

```diff
@@ -184,15 +184,26 @@ module Sysfs = struct
       close_out outchan ;
       raise (Network_error (Write_error file))
 
+  (* device types are defined in linux/if_arp.h *)
+  let is_ether_device name =
+    match int_of_string (read_one_line (getpath name "type")) with
+    | 1 ->
+        true
+    | _ ->
+        false
+    | exception _ ->
+        false
+
   let is_physical name =
     try
       let devpath = getpath name "device" in
       let driver_link = Unix.readlink (devpath ^ "/driver") in
       (* filter out symlinks under device/driver which look like
          /../../../devices/xen-backend/vif- *)
-      not
-        (List.mem "xen-backend"
-           (Astring.String.cuts ~empty:false ~sep:"/" driver_link))
+      is_ether_device name
```

Comment:
I think that this is not quite right, because they _are_ physical devices. It's just that we don't want so see them. I think it is better to filter them out in `list` above.

Reply:
Filtering in `Interface.get_all` may be better, otherwise all internal `Sysfs.list` calls will trigger an additional read for each interface.

Reply:
`Sysfs.list` does not call `is_physical`, so I don't understand this point.

Reply:
The point is to exclude non-ethernet devices from this call: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_pif.ml#L354

Reply:
```
  let devices =
    List.filter
      (fun name -> Net.Interface.is_physical dbg name)
      (Net.Interface.get_all dbg ())
```
This is the code you link to - and it uses the modified `is_physical` to select only devices that we can handle. Why is modifying `get_all` the better solution?

Reply:
We could also expose `is_ether_device` via the RPC mechanism and call it from here - which would make sense to me as well.

Reply:
Actually the problem I have encountered is xen trying to set MAC on a non-eth port on VF, is a VF "physical" ?

Reply:
Because I don't think it is right to make `is_physical` return `false` for devices that are physical. That function is not related to what we are trying to do, apart from the fact that it happens to already be used to filter out devices.

What we really want is: `all devices |> filter ethernet |> filter physical`. You can combine the first two, because we never want to consider non-ethernet devices. We do work with non-physical ethernet devices in other places.

Reply:
Exposing `is_ether_device` also works yes.